c10/core/Stream.h

```diff
@@ -119,21 +119,32 @@ class Stream final {
     // that the bitmasking code below is updated accordingly!
     static_assert(sizeof(DeviceType) == 1, "DeviceType is not 8-bit");
     static_assert(sizeof(DeviceIndex) == 1, "DeviceIndex is not 8-bit");
-    static_assert(sizeof(StreamId) == 4, "DeviceIndex is not 32-bit");
+    static_assert(sizeof(StreamId) == 8, "DeviceIndex is not 64-bit");
     // Concat these together into a 64-bit integer
     // See Note [Hazard when concatenating signed integers]
     uint64_t bits = static_cast<uint64_t>(static_cast<uint8_t>(device_type()))
-            << 48 |
-        static_cast<uint64_t>(static_cast<uint8_t>(device_index())) << 32 |
-        static_cast<uint64_t>(static_cast<uint32_t>(id()));
+            << 56 |
+        static_cast<uint64_t>(static_cast<uint8_t>(device_index())) << 48 |
+        // Remove the sign extension part of the 64-bit address because
+        // the id might be used to hold a pointer.
+        (static_cast<uint64_t>(id()) & ((1ULL << 48) - 1));
+    TORCH_INTERNAL_ASSERT(
+        static_cast<DeviceIndex>((bits >> 48) & 0xFFFFull) == device_index(),
```

Comment:
why is device_index masked with FFFF? It's only 8 bits (49 to 56) now, so the mask should be FF? (It used to be FFFF, when device_index was from bits 33 to 48).